internal/kafka/internal/services/kafka.go

```diff
@@ -400,20 +402,63 @@ func (k *kafkaService) RegisterKafkaJob(kafkaRequest *dbapi.KafkaRequest) *error
 		return errors.TooManyKafkaInstancesReached(fmt.Sprintf("region %s cannot accept instance type: %s at this moment", kafkaRequest.Region, kafkaRequest.InstanceType))
 	}
 
-	if !k.dataplaneClusterConfig.IsDataPlaneAutoScalingEnabled() {
-		cluster, e := k.clusterPlacementStrategy.FindCluster(kafkaRequest)
-		if e != nil || cluster == nil {
-			msg := fmt.Sprintf("no available cluster found for Kafka instance type '%s' in region '%s'", kafkaRequest.InstanceType, kafkaRequest.Region)
-			if e != nil {
-				logger.Logger.Error(fmt.Errorf("%s:%w", msg, e))
-			} else {
-				logger.Logger.Infof(msg)
-			}
+	// enterprise kafkas have their cluster ID assigned when a request is validated in the handler
+	if kafkaRequest.ClusterID != "" {
+		cluster, err := k.clusterService.FindClusterByID(kafkaRequest.ClusterID)
+		if err != nil || cluster == nil {
+			return errors.GeneralError("failed to get cluster with id: %s", kafkaRequest.ClusterID)
+		}
+
+		if cluster.Status != api.ClusterStatus(statusReady.String()) {
+			return errors.BadRequest("cluster with id: %s is not ready to accept kafkas", kafkaRequest.ClusterID)
+		}
+
+		if cluster.OrganizationID != kafkaRequest.OrganisationId {
+			return errors.Unauthorized("user with organization ID: '%s' is unauthorized to create kafka on a cluster belonging to different organization", kafkaRequest.OrganisationId)
+		}
 
-			return errors.TooManyKafkaInstancesReached(fmt.Sprintf("region %s cannot accept instance type: %s at this moment", kafkaRequest.Region, kafkaRequest.InstanceType))
+		instanceSize, getInstanceSizeErr := k.kafkaConfig.GetKafkaInstanceSize(kafkaRequest.InstanceType, kafkaRequest.SizeId)
+		if getInstanceSizeErr != nil {
+			return errors.GeneralError("failed to get kafka instance size: %s", kafkaRequest.SizeId)
 		}
+		capacityInfo := cluster.RetrieveDynamicCapacityInfo()
 
-		kafkaRequest.ClusterID = cluster.ClusterID
+		sizeCountOfCluster, e := k.GetKafkaSizeCountOfEnterpriseCluster(kafkaRequest.ClusterID)
+
+		if e != nil {
+			return errors.GeneralError("failed to establish remaining capacity of cluster with id: %s", kafkaRequest.ClusterID)
+		}
+
+		capacityUsed := int32(0)
+
+		for _, sizeCount := range sizeCountOfCluster {
+			instSize, err := k.kafkaConfig.GetKafkaInstanceSize(types.STANDARD.String(), sizeCount.SizeId)
+			if err != nil {
+				return errors.GeneralError("failed to get kafka instance size: %s", kafkaRequest.SizeId)
+			}
+			capacityUsed = capacityUsed + int32(instSize.CapacityConsumed)*sizeCount.Count
+		}
+		clusterCapacityUsed := int(capacityInfo[kafkaRequest.InstanceType].MaxNodes - capacityUsed)
+		if clusterCapacityUsed < instanceSize.CapacityConsumed {
+			return errors.TooManyKafkaInstancesReached("no available space left on cluster with id: %s for kafka of size: %s", kafkaRequest.ClusterID, kafkaRequest.SizeId)
+		}
+		kafkaRequest.DesiredKafkaBillingModel = "enterprise"
```

Comment:
let's not use a hardcoded billing mode here.

it'll be good to extract somewhere

Reply:
I am aware of this. I see it as bigger refactor though to create constants for all possible billing models. currently we use hardcoded strings everywhere. would you agree to create a follow up Jira to make billing models constant and use them across the codebase?

Reply:
added constant

Reply:
The PR alone will be using the "enterprise" constant at least twice. 
We could just start by moving it onto the https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/blob/ba3f5f14f64443a12a20a2df6d38e045b3637054/internal/kafka/constants/kafka.go constants file and we'll see what can be done about the other billing modes in separate PRs